### ISSUE TITLE
Handling 412 Error on Update

### ISSUE THREAD

**Issue by Arknev** [pnp/pnpjs#248]
### Category
- [ ] Enhancement
- [ ] Bug
- [X] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [1.2.1]

### Expected / Desired Behavior / Question
I'm building a PnPJS/Office Fabric WebPart to handle of all our custom forms going forward. This WebPart will include the Edit Form. Our current JSLink/JSOM custom forms handle overlapping saves OTB. In the interest of maintaining functionality, I am including this in my WebPart. Using the item update from PnPJS I have been able to use etag to return an error. My question is how to handle the error from the item update in a reliable or the intended way. I'm expecting the error status to be available but I can't find a good way to get the value.

### Observed Behavior

I am using this code to send an item update
`
ItemList.items.getById(listItemId).update(itemProperties, this.props.CurrentItemJSON['odata.etag'])
.then(UpdateResponse => {
  console.log('Success');
  console.log(UpdateResponse);
})
.catch(ErrorResponse => {
  console.log('ErrorResponse');
  console.error(ErrorResponse);
});
`

I'm getting this in the console when I trigger an overlapping save issue
> POST https://SITE/_api/web/lists/getByTitle('fieldTesting1')/items(1) 412
> 
> Error: Error making HttpClient request in queryable: [412]  ::> {"responseBody":{"odata.error":{"code":"-1, Microsoft.SharePoint.Client.ClientServiceException","message":{"lang":"en-US","value":"The request ETag value '\"23\"' does not match the object's ETag value '\"24\"'."}}},"responseHeaders":{}} at parsers.ts:59

I can see the 412 and the reason but I can't find a way to read the error type in my code. I see in the parsers.ts file that the response is formatted before returned.

`
reject(new Error('Error making HttpClient request in queryable: [${r.status}] ${r.statusText} ::> ${JSON.stringify(data)}'));
`

I believe that it is the "r.status" value that I want to be available and then handled in my code.

**Comment by patrick-rodgers** [comment]
An answer would be to regex for the code between the '[' ']' chars. Or just look for [412] using like string.contains. BUT, I get what you are asking. Error doesn't by default provide for a way to attach data.We just recently removed our custom error types to simplify things, but perhaps here one is needed. Let me add it back in so it will have a status and statusText property on the error.

**Comment by patrick-rodgers** [comment]
This change was merged in #246 and adds the HttpRequestError class. You can see [the docs in the dev branch](https://github.com/pnp/pnpjs/blob/dev/packages/odata/docs/parsers.md#odatadefaultparser) for usage. Going to close this as resolved, the change will be published in the next beta tomorrow and release in October. Thanks!

**Comment by russgove** [comment]
Hi, slightly off topic but....I built custom forms with spfx and found you could only use them on classic lists. are you using classic or modern lists?

**Comment by Arknev** [comment]
@russgove I'm still in the development stage so my WebPart has only been tested in the workbench. Although, regardless of the experience, modern or classic, the WebPart should work. Since the list itself is neither modern or classic, just a SP list, your code should work. I'm guessing that you are referring to replacing the default forms. I'm not planning on doing that initially. My WP will be used in our solutions to allow for custom UI and applying business rules. These forms will be loaded from custom views and/or custom buttons.

What I know is that two months ago I completed the built of a SPFx WP to replace a solution I created previously that used 7 OTB classic WPs. The new code included the PeoplePicker, a DetailsList, DefaultButtons, PnPJS functions to add items and two custom views. This WP works in classic pages and modern pages in my Tenant.
